iothub_client/src/iothubtransport_mqtt_common.c

```diff
@@ -2174,6 +2176,62 @@ static STRING_HANDLE buildClientId(const char* device_id, const char* module_id)
     }
 }
 
+static int append_optional_connect_parameters(PMQTTTRANSPORT_HANDLE_DATA transport_data)
+{
+    int result = 0;
+
+    if (!transport_data->isOptionalConnectParameterSet)
+    {
+        // This requires the iothubClientHandle, which sadly the MQTT transport only gets on DoWork, so this code still needs to remain here.
+        // The correct place for this would be in the Create method, but we don't get the client handle there.
+        // Also, when device multiplexing is used, the customer creates the transport directly and explicitly, when the client is still not created.
+        // This will be a major hurdle when we add device multiplexing to MQTT transport.
+
+        STRING_HANDLE clone = NULL;
+        STRING_HANDLE param = NULL;
+        STRING_HANDLE urlEncodedModelId = NULL;
+        const char* dt_model_id;
+        const char* product_info = transport_data->transport_callbacks.prod_info_cb(transport_data->transport_ctx);
+
+        if ((clone = (product_info == NULL) ? STRING_construct_sprintf("%s%%2F%s", CLIENT_DEVICE_TYPE_PREFIX, IOTHUB_SDK_VERSION) : URL_EncodeString(product_info)) == NULL)
+        {
+            LogError("Failed obtaining the product info");
+        }
+        else if (STRING_concat_with_STRING(transport_data->configPassedThroughUsername, clone) != 0)
+        {
+            LogError("Failed concatenating the product info");
+        }           
+        else if ((dt_model_id = transport_data->transport_callbacks.dt_model_id_cb(transport_data->transport_ctx)) == NULL)
+        {
+            LogInfo("Not setting device capability model id");
```

Comment:
>LogInfo("Not setting device capability model id"); [](start = 12, length = 50)

Please remove this LogInfo() - it'll be too verbose in the non-PnP scenarios.

Also having an early exit here with the dt_model_id==NULL is not how we typcially code rest of SDK (where falling into an else() of a long block like this means error) so I worry about it confusion a casual reader.

I think a cleaner way will be 

else if ((dt_model_id = blah()  != NULL)
{
     put all the code for handling dt_model_id into this block.  You'll need to have a final "result=0".
}
else {
  result = 0;
}

Reply:
I have refactor with your recommendation but we are breaking the pattern:
if (error)
{ 
    error handling;
}
else if (error)
{
    error handling;
} ...

also, the result is initialized as 0 anyways.  But I did the assignment as suggested.


---
In reply to: [387346346](https://github.com/Azure/azure-iot-sdk-c/pull/1439#discussion_r387346346) [](ancestors = 387346346)